File: code/ex36.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        // Judge rows
        for (int i = 0; i < 9; i++) {
            Set<Character> set = new HashSet<>();
            for (int j = 0; j < 9; j++) {
                if (!ifCharValid(board[i][j]) || set.contains(board[i][j]))    return false;
                if (board[i][j] != '.')    set.add(board[i][j]);
            }
        }
        
        // Judge columns
        for (int j = 0; j < 9; j++) {
            Set<Character> set = new HashSet<>();
            for (int i = 0; i < 9; i++) {
                if (!ifCharValid(board[i][j]) || set.contains(board[i][j]))    return false;
                if (board[i][j] != '.')    set.add(board[i][j]);
            }
        }
        
        // Judge grids
        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                Set<Character> set = new HashSet<>();
                for (int r = i; r < i + 3; r++) {
                    for (int c = j; c < j + 3; c++) {
                        if (!ifCharValid(board[r][c]) || set.contains(board[r][c]))    return false;
                        if (board[r][c] != '.')    set.add(board[r][c]);
                    }
                }
            }
        }
        
        return true;
    }
    
    private boolean ifCharValid(char c) {
        return (c >= '1' && c <= '9') || c == '.';
    }
}

File: code/ex36.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // Row
        for (int i = 0; i < 9; i++) {
            unordered_set<int> set;
            for (int j = 0; j < 9; j++) {
                if (set.count(board[i][j]) > 0)    return 0;
                if (board[i][j] != '.')    set.insert(board[i][j]);
            }
        }
        
        // Col
        for (int j = 0; j < 9; j++) {
            unordered_set<int> set;
            for (int i = 0; i < 9; i++) {
                if (set.count(board[i][j]) > 0)    return 0;
                if (board[i][j] != '.')    set.insert(board[i][j]);
            }
        }
        
        // Grid
        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                if (!judgeGrid(board, i, j))    return 0;
            }
        }
        
        return 1;
    }
    
private:
    // Judge Grid
    bool judgeGrid(vector<vector<char>>& board, int si, int sj) {
        unordered_set<int> set;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (set.count(board[si + i][sj + j]) > 0)    return 0;
                if (board[si + i][sj + j] != '.')    set.insert(board[si + i][sj + j]);
            }
        }
        return 1;
    }
};